### PR TITLE
[frontend] Slowloading of images

### PR DESCRIPTION
**Pull request by klucsik**
# Changes

Add FeatureFlag Provider to client side.
Add http fault filter to envoy
Add slowloading featureflag, and modify the Product card element to disable caching and add the envoy http-fault header
Example of a slowly loaded image trace:
image

## Merge Requirements

For new features contributions please make sure you have completed the following
essential items:

* [x] `CHANGELOG.md` updated to document new feature additions
* [ ] Appropriate documentation updates in the [docs][]
* [ ] Appropriate Helm chart updates in the [helm-charts][]

Maintainers will not merge until the above have been completed. If you're unsure
which docs need to be changed ping the
[@open-telemetry/demo-approvers](https://github.com/orgs/open-telemetry/teams/demo-approvers).

[docs]: https://opentelemetry.io/docs/demo/
[helm-charts]: https://github.com/open-telemetry/opentelemetry-helm-charts

### CODE REVIEW

File: src/flagd/demo.flagd.json [hunk 1]
```diff
@@ -87,6 +87,16 @@
         "off": 0
       },
       "defaultVariant": "off"
+    },
+    "imageSlowLoad": {
+      "description": "slowLoading images in the frontend",
```
Comment (julianocosta89) [inline src/flagd/demo.flagd.json:92]:
```suggestion
      "description": "slow loading images in the frontend",
```

File: src/frontend/components/ProductCard/ProductCard.tsx [hunk 2]
```diff
@@ -5,11 +5,45 @@ import { CypressFields } from '../../utils/Cypress';
 import { Product } from '../../protos/demo';
 import ProductPrice from '../ProductPrice';
 import * as S from './ProductCard.styled';
+import { useState, useEffect } from 'react';
+import { RequestInfo } from 'undici-types';
+import { useNumberFlagValue, OpenFeature} from '@openfeature/react-sdk';
+import {FlagdWebProvider} from '@openfeature/flagd-web-provider'
```
Comment (julianocosta89) [inline src/frontend/components/ProductCard/ProductCard.tsx:11]: nit:
```suggestion
import { useNumberFlagValue, OpenFeature } from '@openfeature/react-sdk';
import { FlagdWebProvider } from '@openfeature/flagd-web-provider';
```

File: docker-compose.yml [hunk 3]
```diff
@@ -314,6 +316,8 @@ services:
         condition: service_started
       imageprovider:
         condition: service_started
+      flagd:
+        condition: service_started
```
Comment (julianocosta89) [inline docker-compose.yml:320]: Please remove the extra space:

```suggestion
        condition: service_started
```

File: src/frontend/components/ProductList/ProductList.tsx [hunk 4]
```diff
@@ -12,11 +12,11 @@ interface IProps {
...
-    <S.ProductList data-cy={CypressFields.ProductList}>
-      {productList.map(product => (
-        <ProductCard key={product.id} product={product} />
-      ))}
-    </S.ProductList>
+      <S.ProductList data-cy={CypressFields.ProductList}>
+        {productList.map(product => (
+            <ProductCard key={product.id} product={product} />
+        ))}
+      </S.ProductList>
```
Comment (julianocosta89) [inline src/frontend/components/ProductList/ProductList.tsx:19]: Is this change needed?

Comment (klucsik) [reply src/frontend/components/ProductList/ProductList.tsx]: I tried to remove this change, I thought I succeeded :D thanks for the catch. 
(historically I've put the provider around this element, but then the product page would throw an error, so I moved it to root level)

File: docker-compose.yml (hunk 3 above)
Comment (klucsik) [reply docker-compose.yml]: thanks for the catch, fixed

File: src/frontend/components/ProductCard/ProductCard.tsx (hunk 2 above)
Comment (klucsik) [reply src/frontend/components/ProductCard/ProductCard.tsx]: thanks for the catch, fixed

File: src/flagd/demo.flagd.json (hunk 1 above)
Comment (klucsik) [reply src/flagd/demo.flagd.json]: thanks for the catch, fixed

**Review by julianocosta89 — COMMENTED** [review]
When trying to build and run this I'm getting the following error:

```
flagd                    | 2024-04-10T12:45:39.071Z	error	evaluator/fractional.go:30	parse fractional evaluation data: bucketing value not supplied and no targetingKey in context
flagd                    | github.com/open-feature/flagd/core/pkg/evaluator.(*Fractional).Evaluate
flagd                    | 	/src/core/pkg/evaluator/fractional.go:30
flagd                    | github.com/diegoholiveira/jsonlogic/v3.operation
flagd                    | 	/go/pkg/mod/github.com/diegoholiveira/jsonlogic/v3@v3.4.0/operation.go:15
flagd                    | github.com/diegoholiveira/jsonlogic/v3.apply
flagd                    | 	/go/pkg/mod/github.com/diegoholiveira/jsonlogic/v3@v3.4.0/jsonlogic.go:493
flagd                    | github.com/diegoholiveira/jsonlogic/v3.ApplyInterface
flagd                    | 	/go/pkg/mod/github.com/diegoholiveira/jsonlogic/v3@v3.4.0/jsonlogic.go:587
flagd                    | github.com/diegoholiveira/jsonlogic/v3.Apply
flagd                    | 	/go/pkg/mod/github.com/diegoholiveira/jsonlogic/v3@v3.4.0/jsonlogic.go:522
flagd                    | github.com/open-feature/flagd/core/pkg/evaluator.(*JSON).evaluateVariant
flagd                    | 	/src/core/pkg/evaluator/json.go:339
flagd                    | github.com/open-feature/flagd/core/pkg/evaluator.resolve[...]
flagd                    | 	/src/core/pkg/evaluator/json.go:276
flagd                    | github.com/open-feature/flagd/core/pkg/evaluator.(*JSON).ResolveAllValues
flagd                    | 	/src/core/pkg/evaluator/json.go:159
flagd                    | github.com/open-feature/flagd/core/pkg/service/flag-evaluation.(*FlagEvaluationService).ResolveAll
flagd                    | 	/src/core/pkg/service/flag-evaluation/flag_evaluator_v2.go:65
flagd                    | connectrpc.com/connect.NewUnaryHandler[...].func1
flagd                    | 	/go/pkg/mod/connectrpc.com/connect@v1.15.0/handler.go:52
flagd                    | connectrpc.com/connect.NewUnaryHandler[...].func2
flagd                    | 	/go/pkg/mod/connectrpc.com/connect@v1.15.0/handler.go:84
flagd                    | connectrpc.com/connect.(*Handler).ServeHTTP
flagd                    | 	/go/pkg/mod/connectrpc.com/connect@v1.15.0/handler.go:265
flagd                    | buf.build/gen/go/open-feature/flagd/connectrpc/go/flagd/evaluation/v1/evaluationv1connect.NewServiceHandler.func1
flagd                    | 	/go/pkg/mod/buf.build/gen/go/open-feature/flagd/connectrpc/go@v1.15.0-20240215170432-1e611e2999cc.1/flagd/evaluation/v1/evaluationv1connect/evaluation.connect.go:245
flagd                    | net/http.HandlerFunc.ServeHTTP
flagd                    | 	/usr/local/go/src/net/http/server.go:2166
flagd                    | github.com/open-feature/flagd/core/pkg/service/flag-evaluation.bufSwitchHandler.ServeHTTP
flagd                    | 	/src/core/pkg/service/flag-evaluation/connect_service.go:51
flagd                    | github.com/open-feature/flagd/core/pkg/service/flag-evaluation.(*ConnectService).setupServer.(*ConnectService).AddMiddleware.Middleware.Handler.func1.1
flagd                    | 	/src/core/pkg/service/middleware/metrics/http_metrics.go:105
flagd                    | github.com/open-feature/flagd/core/pkg/service/middleware/metrics.Middleware.Measure
flagd                    | 	/src/core/pkg/service/middleware/metrics/http_metrics.go:90
flagd                    | github.com/open-feature/flagd/core/pkg/service/flag-evaluation.(*ConnectService).setupServer.(*ConnectService).AddMiddleware.Middleware.Handler.func1
flagd                    | 	/src/core/pkg/service/middleware/metrics/http_metrics.go:104
flagd                    | net/http.HandlerFunc.ServeHTTP
flagd                    | 	/usr/local/go/src/net/http/server.go:2166
flagd                    | github.com/open-feature/flagd/core/pkg/service/flag-evaluation.(*ConnectService).setupServer.(*ConnectService).AddMiddleware.Middleware.Handler.(*Cors).Handler.func2
flagd                    | 	/go/pkg/mod/github.com/rs/cors@v1.10.1/cors.go:281
flagd                    | net/http.HandlerFunc.ServeHTTP
flagd                    | 	/usr/local/go/src/net/http/server.go:2166
flagd                    | golang.org/x/net/http2/h2c.h2cHandler.ServeHTTP
flagd                    | 	/go/pkg/mod/golang.org/x/net@v0.21.0/http2/h2c/h2c.go:125
flagd                    | net/http.serverHandler.ServeHTTP
flagd                    | 	/usr/local/go/src/net/http/server.go:3137
flagd                    | net/http.(*conn).serve
flagd                    | 	/usr/local/go/src/net/http/server.go:2039
```

**Comment by klucsik** [comment]
> When trying to build and run this I'm getting the following error:
> 
> ```
> flagd                    | 2024-04-10T12:45:39.071Z	error	evaluator/fractional.go:30	parse fractional evaluation data: bucketing value not supplied and no targetingKey in context
> flagd                    | github.com/open-feature/flagd/core/pkg/evaluator.(*Fractional).Evaluate
> flagd                    | 	/src/core/pkg/evaluator/fractional.go:30
> flagd                    | github.com/diegoholiveira/jsonlogic/v3.operation
> flagd                    | 	/go/pkg/mod/github.com/diegoholiveira/jsonlogic/v3@v3.4.0/operation.go:15
> flagd                    | github.com/diegoholiveira/jsonlogic/v3.apply
> flagd                    | 	/go/pkg/mod/github.com/diegoholiveira/jsonlogic/v3@v3.4.0/jsonlogic.go:493
> flagd                    | github.com/diegoholiveira/jsonlogic/v3.ApplyInterface
> flagd                    | 	/go/pkg/mod/github.com/diegoholiveira/jsonlogic/v3@v3.4.0/jsonlogic.go:587
> flagd                    | github.com/diegoholiveira/jsonlogic/v3.Apply
> flagd                    | 	/go/pkg/mod/github.com/diegoholiveira/jsonlogic/v3@v3.4.0/jsonlogic.go:522
> flagd                    | github.com/open-feature/flagd/core/pkg/evaluator.(*JSON).evaluateVariant
> flagd                    | 	/src/core/pkg/evaluator/json.go:339
> flagd                    | github.com/open-feature/flagd/core/pkg/evaluator.resolve[...]
> flagd                    | 	/src/core/pkg/evaluator/json.go:276
> flagd                    | github.com/open-feature/flagd/core/pkg/evaluator.(*JSON).ResolveAllValues
> flagd                    | 	/src/core/pkg/evaluator/json.go:159
> flagd                    | github.com/open-feature/flagd/core/pkg/service/flag-evaluation.(*FlagEvaluationService).ResolveAll
> flagd                    | 	/src/core/pkg/service/flag-evaluation/flag_evaluator_v2.go:65
> flagd                    | connectrpc.com/connect.NewUnaryHandler[...].func1
> flagd                    | 	/go/pkg/mod/connectrpc.com/connect@v1.15.0/handler.go:52
> flagd                    | connectrpc.com/connect.NewUnaryHandler[...].func2
> flagd                    | 	/go/pkg/mod/connectrpc.com/connect@v1.15.0/handler.go:84
> flagd                    | connectrpc.com/connect.(*Handler).ServeHTTP
> flagd                    | 	/go/pkg/mod/connectrpc.com/connect@v1.15.0/handler.go:265
> flagd                    | buf.build/gen/go/open-feature/flagd/connectrpc/go/flagd/evaluation/v1/evaluationv1connect.NewServiceHandler.func1
> flagd                    | 	/go/pkg/mod/buf.build/gen/go/open-feature/flagd/connectrpc/go@v1.15.0-20240215170432-1e611e2999cc.1/flagd/evaluation/v1/evaluationv1connect/evaluation.connect.go:245
> flagd                    | net/http.HandlerFunc.ServeHTTP
> flagd                    | 	/usr/local/go/src/net/http/server.go:2166
> flagd                    | github.com/open-feature/flagd/core/pkg/service/flag-evaluation.bufSwitchHandler.ServeHTTP
> flagd                    | 	/src/core/pkg/service/flag-evaluation/connect_service.go:51
> flagd                    | github.com/open-feature/flagd/core/pkg/service/flag-evaluation.(*ConnectService).setupServer.(*ConnectService).AddMiddleware.Middleware.Handler.func1.1
> flagd                    | 	/src/core/pkg/service/middleware/metrics/http_metrics.go:105
> flagd                    | github.com/open-feature/flagd/core/pkg/service/middleware/metrics.Middleware.Measure
> flagd                    | 	/src/core/pkg/service/middleware/metrics/http_metrics.go:90
> flagd                    | github.com/open-feature/flagd/core/pkg/service/flag-evaluation.(*ConnectService).setupServer.(*ConnectService).AddMiddleware.Middleware.Handler.func1
> flagd                    | 	/src/core/pkg/service/middleware/metrics/http_metrics.go:104
> flagd                    | net/http.HandlerFunc.ServeHTTP
> flagd                    | 	/usr/local/go/src/net/http/server.go:2166
> flagd                    | github.com/open-feature/flagd/core/pkg/service/flag-evaluation.(*ConnectService).setupServer.(*ConnectService).AddMiddleware.Middleware.Handler.(*Cors).Handler.func2
> flagd                    | 	/go/pkg/mod/github.com/rs/cors@v1.10.1/cors.go:281
> flagd                    | net/http.HandlerFunc.ServeHTTP
> flagd                    | 	/usr/local/go/src/net/http/server.go:2166
> flagd                    | golang.org/x/net/http2/h2c.h2cHandler.ServeHTTP
> flagd                    | 	/go/pkg/mod/golang.org/x/net@v0.21.0/http2/h2c/h2c.go:125
> flagd                    | net/http.serverHandler.ServeHTTP
> flagd                    | 	/usr/local/go/src/net/http/server.go:3137
> flagd                    | net/http.(*conn).serve
> flagd                    | 	/usr/local/go/src/net/http/server.go:2039
> ```

Hmmm, only fractional is the adServiceFailure. the session span attribute is missing. I don't think I1m touching those requests in my PR, I'll dig around if I can reproduce the issue

**Comment by klucsik** [comment]
The session var should be setted here: https://github.com/open-telemetry/opentelemetry-demo/blob/6f89f569cbab3e0c2127794e1c2b9b95ce140864/src/adservice/src/main/java/oteldemo/AdService.java#L216
This is very far away from my changes, testing main now

**Comment by klucsik** [comment]
main works fine, but my PR somehow introduces these errors, I`ll investigate more

**Comment by klucsik** [comment]
These changes are working as expected, I find no problem in services behavior. But these error messages are emitted every time the react client makes a request against flagd. My suspect would be the interaction between https://github.com/open-feature/js-sdk and flagd.

**Comment by klucsik** [comment]
The problem will be around this caching: https://github.com/open-feature/js-sdk-contrib/tree/main/libs/providers/flagd-web#caching
The provider requesting *all* featureflags, and does not provide any context, and flagd throws errors on this

**Review by austinlparker — APPROVED** [review]
sgtm - are we blocked on the provider being weird?

**Comment by klucsik** [comment]
at provider side, this https://github.com/open-feature/flagd/pull/1295 solves the issue, we had in logs. 
And as I see, this fix is out in the wild now: https://github.com/open-feature/flagd/releases/tag/core%2Fv0.9.1 and this PR bump flagd version to 0.10.1, so we are not bloked

**Review by julianocosta89 — APPROVED** [review]
